packages/server/src/adapters/lambda/index.ts

```diff
@@ -0,0 +1,134 @@
+  APIGatewayProxyResult,
+  APIGatewayProxyStructuredResultV2,
+} from 'aws-lambda';
+import { resolveHTTPResponse } from '../..';
+import { HTTPHeaders, HTTPRequest } from '../../http/internals/types';
+import type { HTTPResponse } from '../../http/internals/types';
+import { AnyRouter, inferRouterContext } from '../../router';
+import type {
+  APIGatewayEvent,
+  APIGatewayResult,
+  AWSLambdaOptions,
+  LambdaContext,
+} from './utils';
+import {
+  UNKNOWN_PAYLOAD_FORMAT_VERSION,
+  isPayloadV1,
+  isPayloadV2,
+} from './utils';
+
+export type { CreateLambdaContextOptions } from './utils';
+
+function lambdaEventToHTTPRequest(event: APIGatewayEvent): HTTPRequest {
+  const query = new URLSearchParams();
+  for (const [key, value] of Object.entries(
+    event.queryStringParameters ?? {},
+  )) {
+    if (value) {
+      query.append(key, value);
+    }
+  }
+
+  return {
+    method: getHTTPMethod(event),
+    query: query,
+    headers: event.headers,
+    body: event.body,
+  };
+}
+
+function getHTTPMethod(event: APIGatewayEvent) {
+  if (isPayloadV1(event)) {
+    return event.httpMethod;
+  } else if (isPayloadV2(event)) {
+    return event.requestContext.http.method;
+  } else {
+    throw UNKNOWN_PAYLOAD_FORMAT_VERSION;
+  }
+}
+function getPath(event: APIGatewayEvent) {
+  if (isPayloadV1(event)) {
+    return event.path.slice(1);
+  } else if (isPayloadV2(event)) {
+    return event.rawPath.slice(1);
+  } else {
+    throw UNKNOWN_PAYLOAD_FORMAT_VERSION;
+  }
+}
+function transformHeaders(headers: HTTPHeaders): APIGatewayResult['headers'] {
+  const obj: APIGatewayResult['headers'] = {};
+
+  for (const [key, value] of Object.entries(headers)) {
+    if (typeof value === 'undefined') {
+      continue;
+    }
+    obj[key] = Array.isArray(value) ? value.join(',') : value;
+  }
+  return obj;
+}
+function tRPCOutputToAPIGatewayOutput<
+  TEvent extends APIGatewayEvent,
+  TResult extends APIGatewayResult,
+>(event: TEvent, response: HTTPResponse): TResult {
+  if (isPayloadV1(event)) {
+    const resp: APIGatewayProxyResult = {
+      statusCode: response.status,
+      body: response.body ?? '',
+      headers: transformHeaders(response.headers ?? {}),
+    };
+    return resp as TResult;
+  } else if (isPayloadV2(event)) {
+    const resp: APIGatewayProxyStructuredResultV2 = {
+      statusCode: response.status,
+      body: response.body ?? undefined,
+      headers: transformHeaders(response.headers ?? {}),
+    };
+    return resp as TResult;
+  } else {
+    throw UNKNOWN_PAYLOAD_FORMAT_VERSION;
+  }
+}
+
+type APIGWReturn<T> = T extends APIGatewayProxyEvent
+  ? APIGatewayProxyResult
+  : T extends APIGatewayProxyEventV2
+  ? APIGatewayProxyStructuredResultV2
+  : never;
+export function lambdaRequestHandler<
+  TRouter extends AnyRouter,
+  TEvent extends APIGatewayEvent,
+  TResult extends APIGWReturn<TEvent>,
+>(
+  opts: AWSLambdaOptions<TRouter, TEvent>,
+): (event: TEvent, context: LambdaContext) => Promise<TResult> {
```

Comment:
Should there be some documented way for an end-user to say they'll only be using APIGatewayEventV2? What would that look like now? Would it just be a matter of providing the generic typeargs explicitly like 

```ts
export const handler = lambdaRequestHandler<
  typeof appRouter,
  APIGatewayEventV2,
  APIGatewayProxyStructuredResultV2
>({
  ...
})
```
?

Reply:
You raise a good point which I forgot about. Some thoughts:

* Internally, it doesn't matter if V1 or V2 is picked. It can deal with both.
* Ideally, what `TEvent` and `TResult` should be should be inferred from the context. I couldn't find how with my current typescript generic skills. I got half way by inferring the return type based on the second parameter. Could `inferRouterContext<TRouter>["event"]` somehow be used?